cpp/src/arrow/array/concatenate.cc

```diff
@@ -436,6 +438,23 @@ class ConcatenateImpl {
     return Status::OK();
   }
 
+  Status Visit(const RunEndEncodedType& type) {
+    int64_t physical_length = 0;
+    for (auto input : in_) {
```

Comment:
Needn't copy the `shared_ptr`
```suggestion
    for (const auto& input : in_) {
```